File: docs/superpowers/staged/2026-09-14/wave3-drafts/m3-t1-t3/build3/frontend/src/types/trends.ts
/**
 * Trends Types
 *
 * Types for the trend comparison sparklines feature (NEM-5406/5407/5408/5409).
 *
 * The trends API provides time-bucketed event metrics with rolling 24-hour
 * baseline comparisons for dashboard sparkline visualization.
 *
 * @see backend/api/schemas/trends.py - Backend Pydantic schemas
 */

// ============================================================================
// Enums
// ============================================================================

/**
 * Type of trend view.
 */
export type TrendType = 'hourly' | 'daily';

// ============================================================================
// API Response Types (snake_case from backend)
// ============================================================================

/**
 * Backend API response for a single trend metric (snake_case).
 * @internal
 */
export interface BackendTrendMetric {
  /** Array of metric values for each time bucket */
  values: number[];
  /** Rolling 24-hour average baseline */
  baseline: number;
  /** Percentage deviation from baseline */
  deviation_pct: number;
}

/**
 * Backend API response for trends endpoint (snake_case).
 * @internal
 */
export interface BackendTrendsResponse {
  /** Event count per time bucket */
  event_count: BackendTrendMetric;
  /** Average risk score per time bucket */
  avg_risk: BackendTrendMetric;
  /** High-risk event count per time bucket */
  high_risk_count: BackendTrendMetric;
}

// ============================================================================
// Frontend Types (camelCase)
// ============================================================================

/**
 * A single trend metric with baseline comparison.
 */
export interface TrendMetric {
  /** Array of metric values for each time bucket (for sparkline display) */
  values: number[];

  /** Rolling 24-hour average baseline for comparison */
  baseline: number;

  /** Percentage deviation from baseline (positive = above, negative = below) */
  deviationPct: number;
}

/**
 * Trends data for dashboard sparkline visualization.
 */
export interface TrendsData {
  /** Event count per time bucket with baseline comparison */
  eventCount: TrendMetric;

  /** Average risk score per time bucket with baseline comparison */
  avgRisk: TrendMetric;

  /** High-risk event count (>= 70) per time bucket with baseline comparison */
  highRiskCount: TrendMetric;
}

// ============================================================================
// Component Props Types
// ============================================================================

/**
 * Props for the TrendSparklines component.
 */
export interface TrendSparklinesProps {
  /** Trends data, or null if loading/unavailable */
  data: TrendsData | null;

  /** Whether trends are currently loading */
  isLoading?: boolean;

  /** Error if fetch failed */
  error?: Error | null;

  /** Callback to retry loading after error */
  onRetry?: () => void;

  /** Additional CSS classes */
  className?: string;

  /** Render in compact mode (smaller dimensions) */
  compact?: boolean;
}

// ============================================================================
// Hook Types
// ============================================================================

/**
 * Return type for useTrends hook.
 */
export interface UseTrendsResult {
  /** Trends data */
  data: TrendsData | null;

  /** Whether data is being fetched */
  isLoading: boolean;

  /** Error if fetch failed */
  error: Error | null;

  /** Manually trigger a refetch */
  refetch: () => Promise<void>;
}

// ============================================================================
// Transformer Functions
// ============================================================================

/**
 * Transform backend trend metric to frontend format.
 *
 * @param backend - Backend trend metric (snake_case)
 * @returns Frontend trend metric (camelCase)
 */
export function transformTrendMetric(backend: BackendTrendMetric): TrendMetric {
  return {
    values: backend.values,
    baseline: backend.baseline,
    deviationPct: backend.deviation_pct,
  };
}

/**
 * Transform backend trends response to frontend format.
 *
 * @param backend - Backend trends response (snake_case)
 * @returns Frontend trends data (camelCase)
 */
export function transformTrendsResponse(backend: BackendTrendsResponse): TrendsData {
  return {
    eventCount: transformTrendMetric(backend.event_count),
    avgRisk: transformTrendMetric(backend.avg_risk),
    highRiskCount: transformTrendMetric(backend.high_risk_count),
  };
}

// ============================================================================
// Type Guards
// ============================================================================

/**
 * Type guard for TrendMetric objects.
 *
 * @example
 * ```ts
 * if (isTrendMetric(data)) {
 *   console.log(data.values);
 * }
 * ```
 */
export function isTrendMetric(obj: unknown): obj is TrendMetric {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    'values' in obj &&
    'baseline' in obj &&
    'deviationPct' in obj &&
    Array.isArray((obj as TrendMetric).values)
  );
}

/**
 * Type guard for TrendsData objects.
 *
 * @example
 * ```ts
 * if (isTrendsData(data)) {
 *   console.log(data.eventCount);
 * }
 * ```
 */
export function isTrendsData(obj: unknown): obj is TrendsData {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    'eventCount' in obj &&
    'avgRisk' in obj &&
    'highRiskCount' in obj &&
    isTrendMetric((obj as TrendsData).eventCount) &&
    isTrendMetric((obj as TrendsData).avgRisk) &&
    isTrendMetric((obj as TrendsData).highRiskCount)
  );
}

File: docs/superpowers/staged/2026-09-14/wave3-drafts/m3-t1-t3/build3/backend/api/schemas/trends.py
"""Pydantic schemas for trends API endpoints.

These schemas define the response models for the trend comparison sparklines feature,
which provides time-bucketed event metrics with rolling 24-hour baseline comparisons.

The trends API supports:
- Hourly view: 12 x 5-minute buckets showing the last hour
- Daily view: 24 x 1-hour buckets showing the last 24 hours

Each metric includes:
- values: Array of data points for sparkline visualization
- baseline: Rolling 24-hour average for comparison
- deviation_pct: Percentage above/below baseline (positive = above, negative = below)
"""

from pydantic import BaseModel, ConfigDict, Field


class TrendMetric(BaseModel):
    """Schema for a single trend metric with baseline comparison.

    Contains time-bucketed values for sparkline display, plus a baseline
    (rolling 24-hour average) and deviation percentage for comparison.

    Example:
        {
            "values": [5, 8, 3, 6, 10, 4, 7, 9, 2, 5, 6, 8],
            "baseline": 6.0,
            "deviation_pct": 33.3
        }
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "values": [5, 8, 3, 6, 10, 4, 7, 9, 2, 5, 6, 8],
                "baseline": 6.0,
                "deviation_pct": 33.3,
            }
        }
    )

    values: list[float] = Field(
        ...,
        description="Array of metric values for each time bucket (for sparkline display)",
    )
    baseline: float = Field(
        ...,
        description="Rolling 24-hour average baseline for comparison",
        ge=0,
    )
    deviation_pct: float = Field(
        ...,
        description="Percentage deviation from baseline (positive = above, negative = below)",
    )


class TrendsResponse(BaseModel):
    """Schema for the trends API response.

    Contains three metrics for dashboard sparkline visualization:
    - event_count: Number of events per time bucket
    - avg_risk: Average risk score per time bucket
    - high_risk_count: Number of high-risk events (risk_score >= 70) per time bucket

    Each metric includes values array, baseline, and deviation percentage.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "event_count": {
                    "values": [5, 8, 3, 6, 10, 4, 7, 9, 2, 5, 6, 8],
                    "baseline": 6.0,
                    "deviation_pct": 33.3,
                },
                "avg_risk": {
                    "values": [45, 52, 38, 60, 72, 40, 55, 65, 35, 48, 50, 58],
                    "baseline": 50.0,
                    "deviation_pct": 16.0,
                },
                "high_risk_count": {
                    "values": [1, 2, 0, 1, 3, 1, 2, 2, 0, 1, 1, 2],
                    "baseline": 1.3,
                    "deviation_pct": 53.8,
                },
            }
        }
    )

    event_count: TrendMetric = Field(
        ...,
        description="Event count per time bucket with baseline comparison",
    )
    avg_risk: TrendMetric = Field(
        ...,
        description="Average risk score per time bucket with baseline comparison",
    )
    high_risk_count: TrendMetric = Field(
        ...,
        description="High-risk event count (>= 70) per time bucket with baseline comparison",
    )
